File: Python/Data Cleaning/VerifyGranularity.py
# Load packages
import pandas as pd

# Define function
def VerifyGranularity(dataframe,
                      list_of_key_columns,
                      keep_key_column=True):
    # Create key column from list of key columns
    first_col = list_of_key_columns[0]
    dataframe['Dataset Key'] = dataframe[first_col].astype(str)
    
    # If more than 1 column listed, concatenate columns together
    if len(list_of_key_columns) > 1:
        for key_col in list_of_key_columns[1:]:
            dataframe['Dataset Key'] = dataframe['Dataset Key'].astype(str) + " -- " + dataframe[key_col].astype(str)
            
    # Print row count and distinct key of Dataset Key
    print("Number of rows in dataframe:", str(len(dataframe.index)))
    print("Distinct count of Dataset Keys:", str(len(pd.unique(dataframe['Dataset Key']))))
    
    # If Key column not being kept, drop it. Otherwise, move Key column to front of DataFrame
    if keep_key_column != True:
        dataframe = dataframe.drop(columns=['Dataset Key'])
    else:
        first_column = dataframe.pop('Dataset Key')
        dataframe.insert(0, 'Dataset Key', first_column)
        
    # Return dataframe
    return(dataframe)

File: R/Data Cleaning/VerifyGranularity.R
# Load packages
library(dplyr)

# Define function
VerifyGranularity = function(dataframe,
                              list_of_key_columns,
                              add_dataset_key_to_dataframe=TRUE,
                              set_dataset_key_as_row_name=TRUE) {
  # Show number of rows in dataframe
  print(paste("Number of rows in dataframe:", nrow(dataframe)))
  
  # Iterate through list of key columnc and create key column
  for (i in 1:length(list_of_key_columns)) {
    if (i == 1) {
      dataframe[["dataset_key"]] = dataframe[[list_of_key_columns[i]]]
    } else {
      dataframe[["dataset_key"]] = paste(dataframe[["dataset_key"]], 
                                         "--",
                                         dataframe[[list_of_key_columns[i]]])
    }
  }
  
  # Show distinct count of key
  print(paste("Distinct count of dataset keys:", n_distinct(dataframe[["dataset_key"]])))
  
  # Delete dataset_key if user does not want to keep it
  if (add_dataset_key_to_dataframe == FALSE) {
    dataframe[["dataset_key"]] = NULL
  } else {
    if (set_dataset_key_as_row_name == TRUE) {
      row.names(dataframe) = dataframe[["dataset_key"]]
      dataframe[["dataset_key"]] = NULL
    } else {
      # Move dataset key to front of dataframe
      dataframe = select(dataframe, 
                         dataset_key, 
                         everything())
    }
  }
  
  # Return dataframe
  return(dataframe)
}
